rfcs/20250416-gemm-external-zp-precompute/README.md

```diff
@@ -0,0 +1,87 @@
+a novel approach has been proposed which is to dynamically downcast both A and B matrices from a floating
+point type to INT8 — but to avoid losing too much precision the downcasting algorithm should also produce
+zero points.
+
+Of particular interest is the variation where the B matrix, or "weights", which is usually constant, gets
+downcast ahead of time. The problem is that zero points on B mean that A (the "sources") is to be reduced
+by the K dimension to multiply the resulting reductions to the ZPs extracted from B; see formulas below.
+
+Calculating partial K reductions of A within the GEMM kernel would never be efficient; there is, however,
+a way to perform them painlessly while downcasting A. Everything that needs adding to GEMM to use them is
+an argument identifier to take that reduction buffer from the user.
+
+The general case, trivial but impractical:
+
+```math
+C_{m,n}=\sum_{k=0}^{K-1}{A_{m,k}(B_{k,n}-Z_B(k))}=\sum_{k=0}^{K-1}{A_{m,k}B_{k,n}-\sum_{k=0}^{K-1}A_{m,k}Z_B(k)}
+```
+
+The grouped B zero point case:
+
+```math
+C_{m,n}=\sum_{g=0}^{G-1}\sum_{k={K\over{G}}g}^{{K\over{G}}(g+1)-1}{A_{m,k}(B_{k,n}-Z_B(g))}=\sum_{k=0}^{K-1}{A_{m,k}B_{k,n}}-\sum_{g=0}^{G-1}Z_B(g)\underbrace{\sum_{k={K\over{G}}g}^{{K\over{G}}(g+1)-1}A_{m,k}}_{R_{m,g}},\;K\mod{G}=0
+```
+
+Where `R` is the matrix of partial K reductions of A, i.e. the reduction buffer introduced above.
+
+## Proposal
+
+A new `DNNL_ARG_ATTR_*` is to be added instead of the long-deprecated `DNNL_ARG_ATTR_OUTPUT_SCALES` along
+with a new convenience arg:
+
+```diff
+ /// Dropout RNG seed value passed via a buffer.
+ #define DNNL_ARG_ATTR_DROPOUT_SEED 511
+ 
+-/// Output scaling factors provided at execution time.
+-/// Deprecated value.
+-#define DNNL_ARG_ATTR_OUTPUT_SCALES 513
++/// The argument marked with this attribute is user-precomputed.
++/// Used exclusively for zero points at the time of writing.
++#define DNNL_ARG_ATTR_USER_PRECOMP 512
+ 
+ /// Starting index for source arguments for primitives that take a variable
+ /// number of source arguments.
+```
+
+```diff
+ 
+ /// Zero points provided at execution time.
+ #define DNNL_ARG_ATTR_ZERO_POINTS 8192
++/// Convenience constant for execution-time user-precomputed zero points.
++#define DNNL_ARG_ATTR_USER_PRECOMP_ZERO_POINTS \
++    (DNNL_ARG_ATTR_USER_PRECOMP | DNNL_ARG_ATTR_ZERO_POINTS)
+ 
+ /// Arguments for fused depthwise convolution.
+ /// See @ref dev_guide_attributes_post_ops_depthwise_fusion
+```
+
+API-wise there is nothing more that should be added, as these 2 arguments are capable of doing everything
+related to accepting any precomputed buffers from the user. All that's needed from the user is to bitwise
+OR either of these constants with the buffer identifier, and pass the result to zero point related oneDNN
+primitive attribute functions (e.g. `dnnl_primitive_attr_set_zero_points`) on PD initialization, and then
```

Comment:
Can you please add an example of how `set_zero_points()` would look like? What should be the argument value? For example now, with regular zero-points, it's just `DNNL_ARG_WEIGHTS`.

Reply:
I expect the precomputed A's to have an ID of `DNNL_ARG_ATTR_USER_PRECOMP | DNNL_ARG_SRC`.

Reply:
Please expand the whole call to `set_zero_points()` because it would definitely have a different semantics.

As a side point - it may likely require a dedicated API to avoid confusion.